src/mbgl/renderer/layers/render_location_component_layer.cpp

```diff
@@ -0,0 +1,895 @@
+        mbgl::gl::bindUniform(simpleShader.u_color, params.errorRadiusColor);
+        mbgl::gl::bindUniform(simpleShader.u_matrix, projectionCircle);
+
+        circleBuffer.upload(circle);
+        MBGL_CHECK_ERROR(glEnableVertexAttribArray(simpleShader.a_pos));
+        MBGL_CHECK_ERROR(glVertexAttribPointer(simpleShader.a_pos, 2, GL_FLOAT, GL_FALSE, 0, nullptr));
+
+        MBGL_CHECK_ERROR(glDrawArrays(GL_TRIANGLE_FAN, 0, circle.size()));
+        if (params.errorRadiusBorderColor.a > 0.0f) {
+            mbgl::gl::bindUniform(simpleShader.u_color, params.errorRadiusBorderColor);
+            MBGL_CHECK_ERROR(glDrawArrays(GL_LINE_STRIP, 1, circle.size() - 1)); // line width: 1 by default;
+        }
+        MBGL_CHECK_ERROR(glDisableVertexAttribArray(simpleShader.a_pos));
+        circleBuffer.detach();
+        simpleShader.detach();
+    }
+
+    void drawQuad(Buffer& buf, std::array<vec2, 4>& data, std::shared_ptr<Texture>& texture) {
+        if (!texture || !texture->isValid()) return;
+        texturedShader.bind();
+        texture->bind(0);
+        glUniform1i(texturedShader.u_image, 0);
+        mbgl::gl::bindUniform(texturedShader.u_matrix, projectionPuck);
+
+        buf.bind();
+        buf.upload(data);
+        MBGL_CHECK_ERROR(glEnableVertexAttribArray(texturedShader.a_pos));
+        MBGL_CHECK_ERROR(glVertexAttribPointer(texturedShader.a_pos, 2, GL_FLOAT, GL_FALSE, 0, nullptr));
+
+        texCoordsBuffer.bind();
+        MBGL_CHECK_ERROR(glEnableVertexAttribArray(texturedShader.a_texCoord));
+        MBGL_CHECK_ERROR(glVertexAttribPointer(texturedShader.a_texCoord, 2, GL_FLOAT, GL_FALSE, 0, nullptr));
+
+        MBGL_CHECK_ERROR(glDrawArrays(GL_TRIANGLE_FAN, 0, 4));
+        texture->detach(); // ToDo: optimize attach and detaches.
+        texCoordsBuffer.detach();
+        texturedShader.detach();
+    }
+
+    void drawShadow() { drawQuad(shadowBuffer, shadowGeometry, texShadow); }
+
+    void drawPuck() { drawQuad(puckBuffer, puckGeometry, texPuck); }
+
+    void drawHat() { drawQuad(hatBuffer, hatGeometry, texPuckHat); }
+
+    static LatLng screenCoordinateToLatLng(const ScreenCoordinate& p,
+                                           const TransformState& s,
+                                           LatLng::WrapMode wrapMode = LatLng::Wrapped) {
+        ScreenCoordinate flippedPoint = p;
+        flippedPoint.y = s.getSize().height - flippedPoint.y;
+        return s.screenCoordinateToLatLng(flippedPoint, wrapMode);
+    }
+
+    void setTextureFromFile(const std::string& imagePath,
+                            std::shared_ptr<Texture>& tex,
+                            const mbgl::style::LocationComponentRenderParameters& params) {
+        if (textures.find(imagePath) == textures.end()) {
+            std::shared_ptr<Texture> tx = std::make_shared<Texture>();
+            if (!imagePath.empty() && params.imageManager)
+                tx->assign(&(params.imageManager->getImage(imagePath)->image));
```

Comment:
Lazy initialize texture once image is set / updated. Taking pointer of a imagemanager's image is unsafe. Client can remove image at any point.